blog/2018/12/22/use-carton-for-your-mojolicious-app-deployment/index.markdown

```diff
@@ -0,0 +1,155 @@
+        requires 'Net::LDAP';
+    };
+
+Cpanfile format can do more (recommended modules, requirements for a specific phase (`configure`, `test`…), using modules not published on CPAN…), but this is a post about Carton: I let you read cpanfile documentation 🙂
+
+Cpanfile can be used by [cpanminus](https://metacpan.org/pod/cpanm) or [Carton](https://metacpan.org/pod/Carton).
+
+Go to the directory containing your `cpanfile` and do:
+
+    cpanm --installdeps .
+
+*Et voilà !*
+
+Note that the modules in `features` has not been installed. You can install them with:
+
+    cpanm --installdeps . --with-feature postgresql
+
+Or, to install all `features` modules, but not the `mysql` one:
+
+    cpanm --installdeps . --with-all-features --without-feature mysql
+
+So, now, we can be sure that we have the good version of our application’s dependencies on the system.
+But what if we host other applications on that system, that have conflicting requirements?
+
+Cpanm is able to install modules in a specific folder (thank you, [local::lib](https://metacpan.org/pod/local::lib)), but wouldn’t it be convenient to install our dependencies in the directory of our application?
+We would always know where our dependencies are.
+
+## Here comes Carton
+
+[Carton](https://metacpan.org/pod/Carton) is Perl module dependency manager. Think `bundler` in Ruby. Think `npm` in Node.js.
+
+Like `npm` does, Carton install the dependencies in the directory of the application.
+
+### Deployment
+
+First, install Carton:
+
+    cpanm Carton
+
+Then, we can install our dependencies with:
+
+    carton install
+
+Our dependencies will be installed in a directory named `local`.
+But there is more: Carton will generate a `cpanfile.snapshot` file, containing the exact version of our dependencies, allowing us to enforce those exact version (ship it with your application).
+
+In our `cpanfile` example, we asked for a Mojolicious version superior or equal to 7.0 and inferior to 8.0.
+Between the installation on our development server and the installation on the production server, some versions may have been published.
+Let’s say that we have Mojolicious 7.77 in our development environment and 7.90 and that something has changed, which leads to problems (for example, the delay helper from [Mojolicious::Plugin::DefaultHelpers](https://mojolicious.org/perldoc/Mojolicious/Plugin/DefaultHelpers) has been [deprecated in 7.78](https://github.com/mojolicious/mojo/blob/47d1369fd11b09af47a76f7f7192985a30ce2409/Changes#L243) and [removed in 7.90](https://github.com/mojolicious/mojo/blob/47d1369fd11b09af47a76f7f7192985a30ce2409/Changes#L150).
+
+Both 7.77 and 7.90 versions are in our range, but our application does not work on the production server… we need to make the production environment as identical as possible as the development one.
+
+For that, since we have a `cpanfile.snapshot` file from our development server, we can do:
+
+    carton install --deployment
```

Comment:
You don't actually say what `--deployment` does. Perhaps add something like

> which installs the exact versions of modules listed in your snapshot.

Reply:
:+1: